---
title: "Results"
date: "Compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    code_folding: hide
  pdf_document: default
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
docache <- TRUE

is_html <- knitr::is_html_output()
knitr::opts_chunk$set(echo=is_html)

this <- system("hostname", TRUE)
if (this == "LAPTOP-IVSPBGCA") { 
	dpath <- "G:/.shortcut-targets-by-id/1mfeEftF_LgRcxOT98CBIaBbYN4ZHkBr_/share/image"
} else if (this == "DESKTOP-M2BA7AA") {
	dpath <- "google drive path"
}

ktab <- function(data, caption) {
    kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width=FALSE) |> 
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))
}

scrolltab <- function(data, caption, width="100%") {
  rownames(data) <- NULL
  kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width = F) |>
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))  |> 
    kableExtra::scroll_box(width=width, height="375px")
}
```


```{r fileread, include=FALSE, cache=docache}
ordnr <- "DMz23-8837_DMz23-8838" #DCas23-7954"
crop <- substr(ordnr, 2, 3)
crp <- substr(ordnr, 2, 3)

f1 <- list.files(file.path(dpath, "input"), 
              pattern=paste0(ordnr, "_SNP.csv"), recursive=TRUE, full=TRUE)
stopifnot(length(f1) == 1)

dirn <- dirname(f1)
country <- substr(dirn, nchar(dirn)-2, nchar(dirn))
match_field <- if (country == "Ethiopia") "target.id" else  "sample" 

inf <- read.csv(gsub("SNP", "variety-info", f1))
dart1 <- matchpoint::read_dart(f1)

#f2 <- gsub("_SNP.csv$", "_SNP_2row.csv", f1)
#fc <- gsub("_SNP.csv$", "_Counts.csv", f1)
#dart2 <- matchpoint::read_dart(f2)
#dartc <- matchpoint::read_dart(fc)
#cns <- colnames(dart1$snp)[-1]


fxl <- gsub("_SNP.csv", "_match_IBS.xlsx", f1)
fxl <- gsub("input", "output/IBS/match", fxl)
isCDS <- grepl("CDS", fxl)
dibs <- readxl::read_excel(fxl, "distance") |> data.frame(check.names=FALSE)
rownames(dibs) <- dibs[,1]
dibs <- dibs[,-1]
fld <- colnames(dibs) %in% inf[[match_field]][!inf$reference]
fdibs <- dibs[fld, fld]

bibs <- readxl::read_excel(fxl, "best_match") |> data.frame()
bibs <- na.omit(bibs)

### compute 
snp1 <- dart1$snp[,-1]
nr <- nrow(snp1)
nc <- ncol(snp1)

# remove snps with many missing values
mr <- rowSums(is.na(snp1)) / nc
msnp <- sum(mr > .5)
msnpp <- round(100 * msnp / length(mr), 1)
snp1 <- snp1[mr <= 0.5, ]
nr1 <- nrow(snp1)

i <- apply(snp1, 1, \(i) length(na.omit(unique(i))))
snp1 <- snp1[i>1,]
sd <- as.data.frame(table(i))
sd[,1] <- as.integer(as.character(sd[,1]))
sinf <- ifelse(nrow(snp1) < nr1, paste(nrow(snp1), "of the "), "All the ")
sinfp <- round(100 * nrow(snp1) / nrow(dart1$snp), 1)
Markers <- c("No variation", "Two variants found", "Three variants found")[sd[,1]]
sd2 <- data.frame(Markers, Count=sd$Freq)

# remove genotypes with many missing values
mrg <- colSums(is.na(snp1)) / nrow(snp1)
mg <- sum(mrg > .5)
mgp <- round(100 * mg / length(mrg), 1)
snp1 <- snp1[, mrg <= 0.5]
m <- match(colnames(snp1), inf[[match_field]])
nref <- sum(inf$reference[m])
nfld <- sum(!inf$reference[m])
sref <- colnames(snp1) %in% inf[[match_field]][inf$reference]

mgs <- ifelse(mg == 0, "No", mg)
method <- if (isCDS) "CDS" else "IBS"
pdftxt <- ifelse(is_html, paste0('<a href="', country, "_", crop, "_results_", method, '.pdf">pdf version</a>'), "")
GE <- ifelse(is_html, "â‰¥", ">=")

get_bm <- function(ibs, n) {
  bm <- as.data.frame(table(ibs$variety))
  bm <- bm[rev(order(bm$Freq)), ]
  bm$perc <- round(100 * bm$Freq / nrow(bibs), 1)
  colnames(bm) <- c("variety", "number of matches", "% of samples")
  bm[,1] <- as.character((bm[,1]))
  add <- bm[1, ]
  add[1,1] <- "not matched"
  add[1,2] <- n - sum(bm[,2])
  add[1,3] <- 100 * (1 - sum(bm[,2]) / n)
  bm <- rbind(bm, add)
  bm[,3] <- round(bm[,3], 1)
  bm
}
```

`r pdftxt`

### Introduction

This page describes the IBS results for `r crop` in `r country` (order `r ordnr`). 


### Matches

```{r gatherref}
ibs8 <- bibs[bibs$IBS > 0.8,]

if (isCDS) {
	thr <- data.frame(
		crp=c("Ri", "Er", "Co", "Mz", "Cp", "Ca"),
		thr=c(.05,  .06,  .06,  .055, .04,  .135)
	)
} else {
	thr <- data.frame(
		crp=c("Ri", "Er", "Co", "Mz", "Cp", "Ca"),
		thr=c(.05,  .07,  .03,   .08,  .05,  .1)
	)
}	

crpth <- 1 - thr$thr[thr$crp==crp]
ibst <- bibs[bibs$IBS > crpth,]

bm8 <- get_bm(ibs8, nrow(bibs))
bmt <- get_bm(ibst, nrow(bibs))

p8 <- round(100 * nrow(ibs8) / nrow(bibs), 1)
pt <- round(100 * nrow(ibst) / nrow(bibs), 1)

allvars <- na.omit(inf$variety)
allvars <- allvars[allvars != ""]
```

`r nrow(ibs8)` (`r p8`%) of the field samples were matched to a reference sample with an IBS score >= 0.8. The reference sample had `r length(unique(allvars))` different varieties;`r nrow(bm8)-1` of these varieties were matched to at least one field sample. The most frequently matched variety was `r bm8[1,1]` which was found `r bm8[1,2]` times. 

At the crop specific threshold of `r crpth`, `r nrow(ibst)` (`r pt`%) of the field samples were matched to a reference sample. `r nrow(bmt)-1` varieties were matched to at least one field sample. The most frequently matched variety was `r bmt[1,1]` which was found `r bmt[1,2]` times. 

<div class = "row">
<div class = "col-md-6">
```{r match_high, fig.width=12, fig.height=5}
bm8 <- bm8[rev(order(bm8[,2])), ]
scrolltab(bm8, caption=paste("IBS", GE, "0.8"))
```
</div>
<div class = "col-md-6">
```{r tab2}
bmt <- bmt[rev(order(bmt[,2])), ]
scrolltab(bmt, caption=paste("IBS", GE, crpth))
```
</div></div>

</br>


```{r plotmatch, fig.height=4, fig.width=9}
par(mar=c(4, 4, 0, 3), mfrow=c(1,2))

x <- (1:nrow(bibs))/nrow(bibs)
plot(x, sort(bibs$IBS), ylab="IBS score", xlab="Field samples matched (fraction)", pch=20, col="red", las=1, cex=.5)
grid()

b <- bm8[bm8[,1] != "not matched", ]
plot(b[,2], las=1, cex=.75, col="red", pch=20, ylab="Number of matches", xlab="Variety")
b <- bmt[bmt[,1] != "not matched", ]
points(1:nrow(b), b[,2], cex=.75, col="blue", pch=20, )
legend("topright", legend=c(paste("IBS", GE, "0.8"), paste("IBS", GE, crpth)), col=c("red", "blue"), pch=20)
       
```
</br>
</br>

### Assignment 

Number of groups (varieties) in the field by IBS dissimilarity threshold.

```{r nvars, fig.width=4.5, fig.height=4}
par(mar=c(4, 4, 0, 1))
cuts <- seq(0.01, 0.2, 0.01)
groupn <- sapply(cuts, \(i) matchpoint:::ngroups.dm(fdibs, i))
plot(c(0, cuts), c(ncol(fdibs), groupn), las=1, ylab="Number of groups", xlab="IBS dissimilarity threshold", type="b", col="red", lwd=2)
grid()
```


```{r assign, fig.width=6, fig.height=25}
cut <- 1-crpth
dm <- matchpoint:::cut.dm(fdibs, cut)
hc <- hclust(dist(dm[[1]])) |> as.dendrogram()
k <- stats::order.dendrogram(hc)
g <- merge(dm[[2]], ibst[, c("field_id", "variety")], by.x="labels", by.y="field_id", all.x=TRUE)
ug <- unique(g$group)
x <- lapply(ug, \(i) {
  v <- g$variety[g$group==i]
  n <- length(v)
  n2 <- sum(is.na(v))
  v <- paste0(sort(unique(v)), collapse = "; ")
  v <- gsub("^; ", "", v)
  out <- paste0(c(n, v), collapse=", ")
  data.frame(group=i, n=n, lab=gsub(", $", "", out))
}) 
x <- do.call(rbind, x)

i <- match(k, x$group)
x <- x[i,]
#hc <- dendextend::set(hc, "labels", x$lab)
#hc <- dendextend::set(hc, "labels_cex", .4)

#par(mar=c(4,0,0,8))
#plot(dendextend::highlight_branches_lwd(hc), horiz=TRUE, nodePar=list(cex=.1), cex=.4)
#abline(v=1-crpth, col="red")
```


### Sample size 

We had data for `r nr` SNPs and `r nc` samples (`r sum(inf$reference)` references and `r sum(!inf$reference)` field samples). `r msnp` markers (`r msnpp`%) had a missing rate that was higher than 0.5. After removing these, we had `r nr1` SNPs left. `r sinf` remaining markers were informative in the sense that they showed variation between genotypes leaving us with `r sinfp`% of the original number of markers. `r mgs` genotypes (`r mgp`%) had a missing rate that was higher than 0.5. After removing these, we had `r ncol(snp1)` genotypes left (`r nref` references and `r nfld` field samples).

<div class = "row">
<div class = "col-md-7">
```{r missingfig, fig.width=6, fig.height=2.5}
par(mfrow=c(1,2), mar=c(4,4,1,0))
plot(sort(mr), las=1, cex=.5, col="blue", ylab="Fraction missing data", xlab="Markers", ylim=c(0,1), cex.axis=.8)
grid()
par(mar=c(4,1,1,3))
plot(sort(mrg), las=1, cex=.5, col="red", ylab="", ylim=c(0,1), xlab="Genotypes", yaxt = "n", cex.axis=.8)
grid()
```
</div>
<div class = "col-md-3">
```{r inftab}
ktab(sd2, "Variation across genotypes in SNP markers")
```
</div></div>


### Full dendogram

```{r dendofull, fig.width=12, fig.height=nrow(bibs)/12}
j <- match(colnames(dibs), ibs8$field_id)
vars <- ibs8$variety[j]
score <- ibs8$IBS[j] |> round(2)
k <- is.na(score)
score[k] <- ""
score[!k] <- paste0("  [", score[!k], "]")

infr <- inf[inf$reference, ]
#j <- match(labs, infr[[match_field]])
j <- match(colnames(dibs), infr[[match_field]])
refs <- infr$variety[j]
isref <- !is.na(refs)
vars[isref] <- refs[isref]
vars[is.na(vars)] <- ""

vars[isref] <- paste0("-- ref -- ", vars[isref])
vars <- paste0(vars, score)

ds <- as.matrix(dibs)
dimnames(ds) <- list(vars, vars)
dc <- matchpoint::group_dend(ds, score)

par(mar=c(0, 3, 1, 7))
plot(dendextend::highlight_branches_lwd(dc), horiz=TRUE, nodePar=list(cex=.1), cex=.6)
abline(v=0.2, col="red")
```

