---
title: "Results"
date: "Compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    code_folding: hide
  pdf_document: default
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
docache <- TRUE

is_html <- knitr::is_html_output()
knitr::opts_chunk$set(echo=is_html)

this <- system("hostname", TRUE)
if (this == "LAPTOP-IVSPBGCA") { 
	dpath <- "G:/.shortcut-targets-by-id/1mfeEftF_LgRcxOT98CBIaBbYN4ZHkBr_/share/image"
} else if (this == "DESKTOP-M2BA7AA") {
	dpath <- "google drive path"
}

ktab <- function(data, caption) {
    kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width=FALSE) |> 
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))
}

scrolltab <- function(data, caption, width="100%") {
  rownames(data) <- NULL
  kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width = F) |>
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))  |> 
    kableExtra::scroll_box(width=width, height="375px")
}
```


```{r fileread, include=FALSE, cache=docache}
ordnr <- "DCas23-7954"
crop <- substr(ordnr, 2, 3)
crp <- substr(ordnr, 2, 3)

f1 <- list.files(file.path(dpath, "input"), 
              pattern=paste0(ordnr, "_SNP.csv"), recursive=TRUE, full=TRUE)
stopifnot(length(f1) == 1)

dirn <- dirname(f1)
country <- substr(dirn, nchar(dirn)-2, nchar(dirn))

inf <- read.csv(gsub("SNP", "variety-info", f1))
dart1 <- matchpoint::read_dart(f1)

fxl <- gsub("_SNP.csv", "_DAP.xlsx", f1)

### compute 
snp1 <- dart1$snp[,-1]
nr <- nrow(snp1)
nc <- ncol(snp1)

# remove snps with many missing values
mr <- rowSums(is.na(snp1)) / nc
msnp <- sum(mr > .5)
msnpp <- round(100 * msnp / length(mr), 1)
snp1 <- snp1[mr <= 0.5, ]
nr1 <- nrow(snp1)

i <- apply(snp1, 1, \(i) length(na.omit(unique(i))))
snp1 <- snp1[i>1,]
sd <- as.data.frame(table(i))
sd[,1] <- as.integer(as.character(sd[,1]))
sinf <- ifelse(nrow(snp1) < nr1, paste(nrow(snp1), "of the "), "All the ")
sinfp <- round(100 * nrow(snp1) / nrow(dart1$snp), 1)
Markers <- c("No variation", "Two variants found", "Three variants found")[sd[,1]]
sd2 <- data.frame(Markers, Count=sd$Freq)

# remove genotypes with many missing values
mrg <- colSums(is.na(snp1)) / nrow(snp1)
mg <- sum(mrg > .5)
mgp <- round(100 * mg / length(mrg), 1)
snp1 <- snp1[, mrg <= 0.5]
m <- match(colnames(snp1), inf$sample)
nref <- sum(inf$reference[m])
nfld <- sum(!inf$reference[m])
sref <- colnames(snp1) %in% inf$sample[inf$reference]

mgs <- ifelse(mg == 0, "No", mg)

pdftxt <- ifelse(is_html, paste0('<a href="', country, "_", crop, '_results.pdf">pdf version</a>'), "")
GE <- ifelse(is_html, "â‰¥", ">=")

get_bm <- function(ibs, n) {
  bm <- as.data.frame(table(ibs$variety))
  bm <- bm[rev(order(bm$Freq)), ]
  bm$perc <- round(100 * bm$Freq / nrow(bibs), 1)
  colnames(bm) <- c("variety", "number of matches", "% of samples")
  bm[,1] <- as.character((bm[,1]))
  add <- bm[1, ]
  add[1,1] <- "not matched"
  add[1,2] <- n - sum(bm[,2])
  add[1,3] <- 100 * (1 - sum(bm[,2]) / n)
  bm <- rbind(bm, add)
  bm[,3] <- round(bm[,3], 1)
  bm
}
```

`r pdftxt`

### Introduction

This page describes the DAP results for `r crop` in `r country` (order `r ordnr`). 

COMING SOON!


