---
title: "Describe Reference"
date: "Compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    code_folding: hide
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
docache <- TRUE

is_html <- knitr::is_html_output()
knitr::opts_chunk$set(echo=is_html)
#knitr::opts_chunk$set(echo = TRUE)

this <- system("hostname", TRUE)
if (this == "LAPTOP-IVSPBGCA") { 
	dpath <- "G:/.shortcut-targets-by-id/1mfeEftF_LgRcxOT98CBIaBbYN4ZHkBr_/share/image"
} else if (this == "DESKTOP-M2BA7AA") {
	dpath <- "google drive path"
}

ktab <- function(data, caption) {
    kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width=FALSE) |> 
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))
}

scrolltab <- function(data, caption, width="100%") {
  rownames(data) <- NULL
  kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width = F,  position = "right") |>
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))  |> 
    kableExtra::scroll_box(width=width, height="400px")
}
```


```{r fileread, include=FALSE, cache=docache}
ordnr <- "DRi23-7825"

#"DCas23-7954"
#"DRi23-7955_moreOrders"

crop <- substr(ordnr, 2, 3)

fc <- list.files(file.path(dpath, "input"), 
              pattern=paste0(ordnr, "_Counts.csv"), recursive=TRUE, full=TRUE)
stopifnot(length(fc) == 1)
dirn <- dirname(fc)
country <- substr(dirn, nchar(dirn)-2, nchar(dirn))
dartc <- matchpoint::read_dart(fc)

inf <- read.csv(gsub("_Counts", "_variety-info", fc))
inf <- inf[inf$reference,]

fxl <- gsub("input", "output/IBS", gsub("_Counts.csv", "_IBS.xlsx", fc))
dibs <- readxl::read_excel(fxl, "distance") |> data.frame(check.names=FALSE)
rownames(dibs) <- dibs[,1]
dibs <- dibs[,-1] |> as.matrix()

cn <- which(colnames(dibs) %in% inf$sample)
dst <- dibs[cn, cn]

all <- nrow(inf)
inf <- inf[inf$sample %in% colnames(dibs), ]
removed <- all - nrow(inf)

#? inf$inventory[is.na(inf$inventory)] <- inf$variety[is.na(inf$inventory)]

nvars <- length(unique(inf$variety))
tab0 <- as.data.frame(table(inf$variety))
ndups <- round(mean(tab0$Freq), 1)

tab1 <- as.data.frame(table(inf$variety, inf$inventory))
tab1 <- tab1[tab1$Freq > 0, ]
techdups <- round(mean(tab1$Freq), 2)
tab2 <- table(tab1$Var1)
biodups <- round(mean(tab2), 2)


bibs <- readxl::read_excel(fxl, "best_match") |> data.frame()
bibs <- bibs[!is.na(bibs$IBS), ]

fdap <- gsub("IBS", "DAP", fxl)
bdap <- readxl::read_excel(fdap, "res_summary") |> data.frame()
best <- merge(bibs[, c("field_id", "IBS", "ref_id", "variety")], bdap[, c("field_id", "ref_id", "variety")], by="field_id")
best <- best[best$IBS >= 0.8, ]
colnames(best) <- gsub("\\.x$", ".IBS", colnames(best))
colnames(best) <- gsub("\\.y$", ".DAP", colnames(best))


#hits <- merge(data.frame(table(best[, "variety.x"])), data.frame(table(best[, "variety.y"])), by=1)
pdftxt <- ifelse(is_html, paste0('<a href="', country, "_", crop, '_reference.pdf">pdf version</a>'), "")
```

`r pdftxt`

### Introduction

This document describes the reference samples for `r crop` in `r country` (order `r ordnr`). There were `r sum(inf$reference)` samples `r ifelse(removed > 0, paste0("(after removing ", removed, " sample(s) for having more than 50% missing data)"), "")`, representing `r nvars` varieties. On average, there were 
`r ndups` samples for each reference variety, `r techdups` samples from the same source ("inventory"; "technical duplicates") and `r biodups` samples from different sources ("biological duplicates").  


### Identification problems

There is a number of problems that can be encountered in a reference set. 

1) Some varieties cannot be clearly separated with the SNP data. These were combined into a single variety, with the original names separated by ` # `. Combinations like `NERICA 1 # NERICA 2 # NERICA 3` are abbreviated to `NERICA 1*2*3` 

2) Some replicates of a variety are clearly in the wrong group. These were renamed to the group the belong to (the renamed references get an `*` appended to their name); 

3) There can be multiple separate groups for the same variety that are genetically very distant. These are identified by appended labels `_a`, `_b`, `_c` etc. These suffixes are dropped if they are combined as under point 1. 

Below is a dendrogram based on the IBS distance and complete linkage clustering. The unit distance is the number of mutations that separate each reference. The labels are formatted as follows: `variety (sample_ID / inventory_ID) [n IBS matches|n DAP matches]`. 
The matches and inventory_ID are omitted if there were none.

The blue and red lines show the cut offs used when making new groups based on the distance matrix (not on this particular dendrogram representation). If a variety has sub-groups that are further apart than the cut-off indicated by the blue line, they are split; unless they are nearest neighbors to each other, in which case they are kept together. Varieties below the red line are lumped together.


```{r plotdendo2, fig.width=10, fig.height=nrow(inf)/8}
dcn <- colnames(dst)

b_ibs <- as.data.frame(table(best$ref_id.IBS))
b_dap <- as.data.frame(table(best$ref_id.DAP))
k <- match(colnames(dst), b_ibs[,1])
h_ibs <- b_ibs[k, 2]
k <- match(colnames(dst), b_ibs[,1])

h_dap <- b_dap[k, 2]
h <- paste0(h_ibs, "|", h_dap)
h <- gsub("NA", "0", h)
h <- paste0("  [", h, "]")
hits <- gsub("  \\[0\\|0]", "", h)

j <- match(colnames(dst), inf$sample)
variety <- inf$variety[j]

if (!all(is.na(inf$inventory[j]))) {
	add <- paste0("  (", inf$sample[j], " / ", inf$inventory[j],") ", hits)
} else {
	add <- paste0("  (", inf$sample[j], ") ", hits)
}

dimnames(dst) <- list(variety, variety)
#attr(dst, "Labels") <- variety
dc <- matchpoint::group_dend(dst, add)
par(mar=c(3, 0, 0, 35))
plot(dendextend::highlight_branches_lwd(dc), horiz=TRUE, nodePar=list(cex=.1), cex=.6)

maxcut <- ifelse(crop=="rice", .5, .3)

v1 <- matchpoint::self_dist(dst)

# split and lump
splum <- matchpoint:::split_lump(dst, .01, .05) 

k <- stats::order.dendrogram(dc)
#var1 <- variety[k]
var2 <- splum$new[k]
changed <- variety[k] != var2
var2[!changed] <- ""

set.seed(1)
rc2 <- viridis::turbo(length(unique(var2)))
#cols2 <- rep(NA, length(var2))
ref2 <- as.integer(as.factor(var2))
cols2 <- rc2[ref2]

#plot(cbind(0, 1:length(var1)), type="n", axes=FALSE)
pd <- diff(par("usr")[1:2]) 

text(cbind(pd, 1:length(var2)), labels=var2, col=cols2, pos=4, xpd=TRUE, cex=0.55)

text(0, length(var2)+2, labels="Original name", pos=4, xpd=TRUE, cex=1)
text(pd, length(var2)+2, labels="Fixed name", pos=4, xpd=TRUE, cex=1)

lines(cbind(splum$lump, c(0, length(variety))), col=rgb(1,0,0,0.5), lwd=1)
text(splum$lump, 0, labels=round(splum$lump), adj=0, xpd=TRUE, cex=.5, col="red")
lines(cbind(splum$split, c(0, length(variety))), col=rgb(0,0,1,0.5), lwd=2)
text(splum$split, 0, labels=round(splum$split), adj=1, xpd=TRUE, cex=.5, col="blue")


refs <- intersect(dcn, colnames(dibs))
i <- dcn %in% refs
j <- colnames(dibs) %in% refs
x <- cbind(expand.grid(refs, refs), dst=as.vector(dst[i,i]), ibs=round(1000*as.vector(dibs[j,j])))
x <- x[x[,1] != x[,2], ]
x[,1:2] <- t(apply(x[,1:2], 1, sort))
x <- unique(x)
x$ibs = x$ibs/1000
colnames(x)[1:2] <- c("id1", "id2")

inf2 <- merge(inf, data.frame(sample=dcn, to=splum$new), by="sample")
x <- merge(x, inf2[, c("sample", "inventory", "to")], by=1)
colnames(x)[5:6] <- c("inv1", "var1")
x <- merge(x, inf2[, c("sample", "inventory", "to")], by.x=2, by.y=1)
colnames(x)[7:8] <- c("inv2", "var2")
x <- x[x$var1 == x$var2,]

trep_ibs <- median(x$ibs[x$inv1 == x$inv2])
brep_ibs <- median(x$ibs[x$inv1 != x$inv2])
trep_dst <- median(x$dst[x$inv1 == x$inv2])
brep_dst <- median(x$dst[x$inv1 != x$inv2])

scores <- round(c(trep_dst, brep_dst, trep_ibs, brep_ibs), 3)

dimnames(dst) <- list(splum$new, splum$new)
v <- matchpoint::self_dist(dst)
```

The table below shows the changes made in the labels. 

```{r changtab}
changed <- merge(inf[, c("sample", "variety")], 
          data.frame(sample=dcn, to=colnames(dst)), by="sample")
colnames(changed)[2] <- "from"
changed <- changed[changed$from != changed$to,]
scrolltab(changed, "", width="50%")
```

</br>
</br>

### Distance within and between varieties

#### within

After the cleaning suggested above, the median average IBS distance between samples of the same variety was `r round(median(v$value), 1)`; `r scores[1]` for technical replicates and `r scores[2]` for biological replicates. 

```{r bplot, fig.width=3.5, fig.height=5}
vv <- rbind(data.frame(id="before cleaning", v=v1$value), data.frame(id="after cleaning", v=v$value))
par(mar=c(2,6,1,0))
boxplot(v~id, data=vv, ylab="Dissimilarity within varieties\n(IBS)", las=1, col=c("red", "blue"), xlab="", xpd=TRUE)

nn <- matchpoint::nngb_dist(dst)$value
qnn <- round(quantile(nn), 1)
noth <- matchpoint::other_dist(dst)$value
```

#### between

The lowest observed distance between two varieties was `r qnn[1]`, the median was `r qnn[3]` and the largest distance between a variety and its nearest genetic neighbor in the reference set was `r qnn[5]`

```{r bplot2, fig.width=6, fig.height=3.5}
par(mar=c(0, 6, 3, 2), mfrow=c(1,2))
boxplot(nn, ylab="distance to nearest neighbor\n(IBS)", las=1, col="red", cex=.5)
```

#### ratio

The ratio of the median within and between varieties mutation distance was `r round(qnn[3]/median(noth), 2)`. 

</br>

### Single genotype?

In the case of diploid clonally propagated crops and inbreeding species, one would expect that, if a sample consisted of a single genotype, the counts for each heterozygous SNP variant would have a ratio of about 1. The "heterozygosity index" is the median score of these ratios for all heterozygous markers of a variety. Samples with less then 20% heterozygous markers are not considered.

```{r purity, fig.width=4, fig.height=4, cache=FALSE}
out <- matchpoint:::purity(dartc, mr=0.5, minhet=0)  
out <- out[colnames(dartc$snp)[-(1:2)] %in% inf$sample] |> sort()
nv <- length(out)

par(mar=c(4,4,1,1))
plot((1:nv)/nv, out, las=1, pch=20, 
     cex=1, lwd=2,
     xlab="fraction of samples", 
     ylab="heterozygosity index",
     col="orange")
grid()

```


