---
title: "Describe SNP data"
date: "Compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    code_folding: hide
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
order <- "DCob22-7521_moreOrders"
knitr::opts_chunk$set(echo = TRUE)

this <- system("hostname", TRUE)
if (this == "LAPTOP-IVSPBGCA") { 
	dpath <- "G:/.shortcut-targets-by-id/1mfeEftF_LgRcxOT98CBIaBbYN4ZHkBr_/share/image"
} else if (this == "DESKTOP-M2BA7AA") {
	dpath <- "google drive path"
}

ktab <- function(data, caption) {
    kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width=FALSE) |> 
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))
}

scrolltab <- function(data, caption) {
kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width = F,  position = "right") |>
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))  |> 
    kableExtra::scroll_box(width="50%", height="400px")
}
```


```{r fileread, include=FALSE, cache=FALSE}
f1 <- list.files(file.path(dpath, "input"), 
              pattern=paste0(order, "_SNP.csv"), recursive=TRUE, full=TRUE)
stopifnot(length(f1) == 1)
f2 <- gsub("_SNP.csv$", "_SNP_2row.csv", f1)

dirn <- dirname(f1)
country <- substr(dirn, nchar(dirn)-2, nchar(dirn))
crop <- substr(order, 2, 3)

inf <- read.csv(gsub("SNP", "variety-info", f1))
x <- matchpoint::read_dart(f1)
x2 <- matchpoint::read_dart(f2)

fxl <- gsub("_SNP.csv", "_IBS.xlsx", f1)
fxl <- gsub("input", "output/IBS", fxl)
dibs <- readxl::read_excel(fxl, "distance")
bibs <- readxl::read_excel(fxl, "best_match") |> data.frame()
bibs <- bibs[!is.na(bibs$field_id), ]
#ribs <- readxl::read_excel(fxl, "ref")
```

### Introduction

Describing order: `r order` for `r crop` in `r country`

The raw data had `r nrow(x$snp)` SNPs and `r ncol(x$snp)-1` samples (`r sum(inf$reference)` references and `r sum(!inf$reference)` field samples).  

```{r missing, fig.width=7, fig.height=3.5}
par(mfrow=c(1,2), mar=c(4,4,1,0))

snp <- x$snp[,-c(1:2)]
mr <- rowSums(is.na(snp)) / ncol(snp)
msnp <- sum(mr >.2)
msnpp <- round(100 * msnp / length(mr), 1)
snp <- snp[mr <= 0.2, ]

mrg <- colSums(is.na(snp)) / nrow(snp)

plot(sort(mr), las=1, cex=.5, col="blue", ylab="Fraction missing data", xlab="Markers", ylim=c(0,1), cex.axis=.8)
grid()

par(mar=c(4,1,1,3))
plot(sort(mrg), las=1, cex=.5, col="red", ylab="", ylim=c(0,1), xlab="Genotypes", yaxt = "n", cex.axis=.8)
grid()
```
</br></br>
`r msnp` markers (`r msnpp`%) had a missing rate that was higher than 0.2. After removing these, we had `r nrow(snp)` SNPs left.

```{r markvariation}
i <- apply(snp, 1, \(i) length(na.omit(unique(i))))
sd <- as.data.frame(table(i))
sinf <- nrow(snp) - sd[sd[,1]==1, 2]
sinfp <- round(100 * sinf / nrow(snp), 1)
Markers <- c("No variation", "Two variants found", "Three variants found")[sd[,1]]
sd2 <- data.frame(Markers, Count=sd$Freq)
ktab(sd2, "Variation across genotypes in SNP markers")
```

</br>
Of the `r nrow(x$snp)` remaining markers, `r sinf` (`r sinfp`%) were informative in the sense that they showed variation between genotypes. 
</br>

```{r genotypes, fig.width=5}
mg <- sum(mrg >.2)
mgp <- round(100 * mg / length(mrg), 1)
snp <- snp[, mrg <= 0.2]
m <- match(colnames(snp), inf$sample)
nref <- sum(inf$reference[m])
nfld <- sum(!inf$reference[m])

sref <- colnames(x2$snp) %in% inf$sample[inf$reference]
```

</br>
`r mg` genotypes (`r mgp`%) had a missing rate that was higher than 0.2. After removing these, we have `r ncol(snp)` genotypes left (`r nref` references and `r nfld` field samples).

### Reference sample

```{r gatherref}
ibs <- bibs[bibs$IBS > 0.8,]
bm <- as.data.frame(table(ibs$ref_id))
ref <- x2$snp[, sref]
dst <- matchpoint:::hamming_distance(ref)
hc <- hclust(as.dist(dst))
labs <- hc$labels[hc$order]
j <- match(labs, inf$sample)
k <- match(labs, bm[,1])
hits <- bm[k, 2]
hits[!is.na(hits)] <- paste0(" [", hits[!is.na(hits)], "]")
hits[is.na(hits)] <- ""
variety <- inf$variety[j]
#, "/", inf$inventory[j]
variety2 <- paste0(inf$variety[j], " (", inf$sample[j], ") ", hits)

vars <- na.omit(inf$variety)
vars <- vars[vars != ""]
bm <- bm[rev(order(bm$Freq)), ]
colnames(bm) <- c("variety", "number of matched field samples")
```

`r nrow(ibs)` (`r round(100*nrow(ibs)/nrow(bibs), 1)`%) of the reference samples were matched with IBS. 

The reference sample had `r length(vars)` different varieties, each one represented `r round(length(unique(vars))/length(vars), 1)` times, on average.

`r nrow(bm)` of these varieties were matched to at least one reference sample. The most commonly found variety was `r bm[1,1]` which was matched `r bm[1,2]` times. 


```{r plotmatch, fig.height=6, fig.width=5}
scrolltab(bm, "Number of matched field samples by reference variety")
lab <- as.character(bm[,1])
lab[bm[,2] < 15] <- ""
plot(bm[,2], las=1, cex=.4, col="red", pch=20, ylab="nuber of matches", xlab="order")
text(1:nrow(bm), bm[,2], lab, cex=.5, xpd=TRUE, pos=4, col="blue")
```


Below is a dendrogram based on the Hamming distance and complete linkage clustering. The unit distance is the number of mutations that separate each reference. The labels are formatted as follows: `variety (sample-id) [number of IBS matches]`

```{r plotdendo1, fig.width=6, fig.height=sum(sref)/10}
set.seed(1)
rc <- sample(rainbow(length(unique(variety))))
cols <- rep(NA, length(variety))
ref <- as.integer(as.factor(variety))
cols <- rc[ref]
dc <- as.dendrogram(hc)
dc <- dendextend::set(dc, "labels", variety2)
dc <- dendextend::set(dc, "labels_cex", 0.5)
dc <- dendextend::set(dc, "labels_col", cols)
par(mar=c(4, 1, 1, 10))
plot(dc, horiz=TRUE, nodePar=list(cex=.1))
```

