---
title: "Describe SNP data"
date: "Compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    code_folding: hide
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
ordnr <- "DCas23-7954"
docache <- TRUE
crop <- substr(ordnr, 2, 3)

knitr::opts_chunk$set(echo = TRUE)

this <- system("hostname", TRUE)
if (this == "LAPTOP-IVSPBGCA") { 
	dpath <- "G:/.shortcut-targets-by-id/1mfeEftF_LgRcxOT98CBIaBbYN4ZHkBr_/share/image"
} else if (this == "DESKTOP-M2BA7AA") {
	dpath <- "google drive path"
}

ktab <- function(data, caption) {
    kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width=FALSE) |> 
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))
}

scrolltab <- function(data, caption) {
  rownames(data) <- NULL
  kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width = F,  position = "right") |>
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))  |> 
    kableExtra::scroll_box(width="100%", height="400px")
}
```


```{r fileread, include=FALSE, cache=docache}
f1 <- list.files(file.path(dpath, "input"), 
              pattern=paste0(ordnr, "_SNP.csv"), recursive=TRUE, full=TRUE)
stopifnot(length(f1) == 1)
f2 <- gsub("_SNP.csv$", "_SNP_2row.csv", f1)
fc <- gsub("_SNP.csv$", "_Counts.csv", f1)

dirn <- dirname(f1)
country <- substr(dirn, nchar(dirn)-2, nchar(dirn))

inf <- read.csv(gsub("SNP", "variety-info", f1))
x <- matchpoint::read_dart(f1)
x2 <- matchpoint::read_dart(f2)
xc <- matchpoint::read_dart(fc)

cns <- colnames(x$snp)[-1]

stopifnot(all(colnames(x$snp)[-1] == colnames(x2$snp)[-(1:2)]))

fxl <- gsub("_SNP.csv", "_IBS.xlsx", f1)
fxl <- gsub("input", "output/IBS", fxl)
dibs <- readxl::read_excel(fxl, "distance")
bibs <- readxl::read_excel(fxl, "best_match") |> data.frame()
bibs <- bibs[!is.na(bibs$field_id), ]
```

### Introduction

Describing order: `r ordnr` for `r crop` in `r country`

The raw data had `r nrow(x$snp)` SNPs and `r ncol(x$snp)-1` samples (`r sum(inf$reference)` references and `r sum(!inf$reference)` field samples).  

```{r missing, fig.width=7, fig.height=3.5}
par(mfrow=c(1,2), mar=c(4,4,1,0))

snp <- x$snp[,-c(1:2)]
mr <- rowSums(is.na(snp)) / ncol(snp)
msnp <- sum(mr >.2)
msnpp <- round(100 * msnp / length(mr), 1)
snp <- snp[mr <= 0.2, ]

mrg <- colSums(is.na(snp)) / nrow(snp)

plot(sort(mr), las=1, cex=.5, col="blue", ylab="Fraction missing data", xlab="Markers", ylim=c(0,1), cex.axis=.8)
grid()

par(mar=c(4,1,1,3))
plot(sort(mrg), las=1, cex=.5, col="red", ylab="", ylim=c(0,1), xlab="Genotypes", yaxt = "n", cex.axis=.8)
grid()
```
</br></br>
`r msnp` markers (`r msnpp`%) had a missing rate that was higher than 0.2. After removing these, we had `r nrow(snp)` SNPs left.

```{r markvariation}
i <- apply(snp, 1, \(i) length(na.omit(unique(i))))
nrow1 <- nrow(snp)
snp <- snp[i>1,]

sd <- as.data.frame(table(i))
sd[,1] <- as.integer(as.character(sd[,1]))
sinf <- ifelse(nrow(snp) < nrow1, paste(nrow(snp), "of the "), "All the ")
sinfp <- round(100 * nrow(snp) / nrow(x$snp), 1)
Markers <- c("No variation", "Two variants found", "Three variants found")[sd[,1]]
sd2 <- data.frame(Markers, Count=sd$Freq)
ktab(sd2, "Variation across genotypes in SNP markers")
```

</br>
`r sinf` remaining markers were informative in the sense that they showed variation between genotypes leaving us with (`r sinfp`% ) of the original number of markers. 
</br>

```{r genotypes, fig.width=5}
mg <- sum(mrg >.2)
mgp <- round(100 * mg / length(mrg), 1)
snp <- snp[, mrg <= 0.2]
m <- match(colnames(snp), inf$sample)
nref <- sum(inf$reference[m])
nfld <- sum(!inf$reference[m])
sref <- colnames(snp) %in% inf$sample[inf$reference]

mgs <- ifelse(mg == 0, "No", mg)
```

</br>
`r mgs` genotypes (`r mgp`%) had a missing rate that was higher than 0.2. After removing these, we have `r ncol(snp)` genotypes left (`r nref` references and `r nfld` field samples).

### Single genotype?

In the case of clonally propagated or inbreeding species, one would expect that, if a sample consisted of a single genotype, the counts for each heterozygous SNP variant would have a proportion of 1. The "heterozygosity index" is the median score for all heterozygous markers of a variety. Since our samples consisted of 5 plants, I only SNP pairs with at least 5 counts in total, and with proportions of at least 0.2. Also, samples with less then 25 heterozygous markers are not considered.

```{r purity, fig.width=4, fig.height=4, cache=docache}
p <- as.matrix(xc$snp[,-(1:2)])
i <- colSums(is.na(p)) < (0.2 * nrow(p))
p <- p[,i]
# not hetero
p[p==0] <- NA
# at least 10 heteros
i <- seq(1, nrow(p), 2)
p1 <- p[i,]
p2 <- p[i+1,]
nv <- ncol(p1) 
out <- rep(NA, nv)
for (i in 1:nv) {
    pi <- na.omit(cbind(p1[,i], p2[,i]))
    if (!(isTRUE(nrow(pi) > 25))) next
    pi <- pi[rowSums(pi)> 4, ]
    if (!(isTRUE(nrow(pi) > 25))) next
    pi <- apply(pi, 1, sort) |> t()
    s <- pi[,1] / pi[,2]
    # if s if very small it cannot be meaningful
    out[i] <- median(s[s > 0.2])
}
out <- sort(out)
nv <- length(out)
par(mar=c(4,4,1,1))
plot((1:nv)/nv, out, las=1, pch=20, 
     cex=.5, lwd=2,
     xlab="fraction of samples", 
     ylab="heterozygosity index",
     col="orange")
grid()

```


### Reference sample

```{r gatherref}
ibs <- bibs[bibs$IBS > 0.8,]
bm <- as.data.frame(table(ibs$ref_id))
sref <- colnames(x2$snp) %in% inf$sample[inf$reference]
ref <- x2$snp[, sref]
ref <- matchpoint:::remove_sparse_records(ref)
dst <- matchpoint:::hamming_distance(ref)
#hc <- dendextend::find_dend(as.dist(dst))
hc <- hclust(as.dist(dst))
labs <- hc$labels[hc$order]
j <- match(labs, inf$sample)
k <- match(labs, bm[,1])
hits <- bm[k, 2]
hits[!is.na(hits)] <- paste0(" [", hits[!is.na(hits)], "]")
hits[is.na(hits)] <- ""
variety <- inf$variety[j]
if (any(is.na(inf$sample[j]))) {
	variety2 <- paste0(variety, " (", inf$sample[j], " / ", inf$inventory[j],") ", hits)
} else {
	variety2 <- paste0(variety, " (", inf$sample[j], ") ", hits)
}
vars <- na.omit(inf$variety)
vars <- vars[vars != ""]
bm <- bm[rev(order(bm$Freq)), ]
colnames(bm) <- c("variety", "number of matched field samples")
```

`r nrow(ibs)` (`r round(100*nrow(ibs)/nrow(bibs), 1)`%) of the reference samples were matched with IBS. 

The reference sample had `r length(unique(vars))` different varieties, each one represented `r round(length(vars)/length(unique(vars)), 1)` times, on average.

`r nrow(bm)` of these varieties were matched to at least one reference sample. The most commonly found variety was `r bm[1,1]` which was matched `r bm[1,2]` times. 

<div class = "row">
<div class = "col-md-6">
```{r plotmatch, fig.height=6, fig.width=5}
lab <- as.character(bm[,1])
lab[bm[,2] < 15] <- ""
par(mar=c(4, 4, 0, 1))
plot(bm[,2], las=1, cex=.4, col="red", pch=20, ylab="nuber of matches", xlab="order")
text(1:nrow(bm), bm[,2], lab, cex=.5, xpd=TRUE, pos=4, col="blue")
```
</div>
<div class = "col-md-5">
```{r tabmatch}
scrolltab(bm, "Number of matched field samples by reference variety")
```
</div></div>


Below is a dendrogram based on the Hamming distance and complete linkage clustering. The unit distance is the number of mutations that separate each reference. The labels are formatted as follows: `variety (sample_ID / inventory_ID) [number of IBS matches]` or `variety (sample_ID) [number of IBS matches]` if there was no inventory_ID.

The grey lines show the median, third quartile, `third quartile + 1.5 IQR` and maximum within variety distance, see below.)

```{r plotdendo1, fig.width=6, fig.height=sum(sref)/10}
diag(dst) <- NA
jj <- match(colnames(dst), inf$sample)
var <- inf$variety[jj]
u <- unique(var)
v <- sapply(1:length(u), \(i) {
  j <- which(u[i] == var)
  dv <- dst[j, j]
  mean(dv, na.rm=TRUE)
})
v[!is.finite(v)] <- 0
set.seed(1)
rc <- sample(rainbow(length(unique(variety))))
cols <- rep(NA, length(variety))
ref <- as.integer(as.factor(variety))
cols <- rc[ref]
dc <- as.dendrogram(hc)
dc <- dendextend::set(dc, "labels", variety2)
dc <- dendextend::set(dc, "labels_cex", 0.5)
dc <- dendextend::set(dc, "labels_col", cols)

qth <- quantile(v)
iqr3 <- qth[4] + diff(qth[c(2,4)])

par(mar=c(3, 1, 0, 10))
plot(dendextend::highlight_branches(dc), horiz=TRUE, nodePar=list(cex=.1), cex=.6)
lines(cbind(qth[2], c(0, length(variety))), col=gray(.5, alpha=.25), lwd=.5)
lines(cbind(qth[3], c(0, length(variety))), col=gray(.5, alpha=.25), lwd=1)
lines(cbind(qth[4], c(0, length(variety))), col=gray(0.25, alpha=.25), lwd=2)
lines(cbind(iqr3, c(0, length(variety))), col=gray(0, alpha=.25), lwd=3)
lines(cbind(qth[5], c(0, length(variety))), col=rgb(1,0,0,0.25), lwd=4)
```


The median average Hamming distance between samples of the same variety is `r round(median(v), 1)`% mutations (the number of mutations relative to the number of usable SNPs in the assay). This is before correcting the reference for any obvious inconsistencies.

```{r bplot, fig.width=3, fig.height=3.5}
par(mar=c(4, 6, 1, 0))
fv <- 100*v/nrow(snp)
boxplot(fv, ylab="Dissimilarity within varieties\n(% mutations)", las=1)
```


## Trimming 

```{r plotdendocln, fig.width=6, fig.height=sum(sref)/10}
j <- match(colnames(dst), inf$sample)
vars <- inf$variety[j]
dcl <- matchpoint:::clean_dist_labels(dst, vars, iqr3, nmax=100, verbose=FALSE) 
scrolltab(dcl[dcl$changed, c("sample", "from", "to")], "")

j <- match(labs, dcl$sample)
var1 <- dcl$to[j]
var2 <- gsub(" \\*", "", var1)

set.seed(1)
rc <- sample(rainbow(length(unique(var2))))
cols <- rep(NA, length(var2))
cols <- rc[as.integer(as.factor(var2))]
dc <- dendextend::set(dc, "labels", var1)
dc <- dendextend::set(dc, "labels_col", cols)

par(mar=c(3, 1, 0, 10))
plot(dendextend::highlight_branches(dc), horiz=TRUE, nodePar=list(cex=.1), cex=.6)
```

