---
title: "Describe SNP data"
date: "Compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    code_folding: hide
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
docache <- TRUE

is_html <- knitr::is_html_output()
knitr::opts_chunk$set(echo=is_html)
#knitr::opts_chunk$set(echo = TRUE)

this <- system("hostname", TRUE)
if (this == "LAPTOP-IVSPBGCA") { 
	dpath <- "G:/.shortcut-targets-by-id/1mfeEftF_LgRcxOT98CBIaBbYN4ZHkBr_/share/image"
} else if (this == "DESKTOP-M2BA7AA") {
	dpath <- "google drive path"
}

ktab <- function(data, caption) {
    kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width=FALSE) |> 
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))
}

scrolltab <- function(data, caption, width="100%") {
  rownames(data) <- NULL
  kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width = F,  position = "right") |>
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))  |> 
    kableExtra::scroll_box(width=width, height="400px")
}
```


```{r fileread, include=FALSE, cache=docache}
ordnr <- "DCas23-7954"
crop <- substr(ordnr, 2, 3)

f1 <- list.files(file.path(dpath, "input"), 
              pattern=paste0(ordnr, "_SNP.csv"), recursive=TRUE, full=TRUE)
stopifnot(length(f1) == 1)
f2 <- gsub("_SNP.csv$", "_SNP_2row.csv", f1)
fc <- gsub("_SNP.csv$", "_Counts.csv", f1)

dirn <- dirname(f1)
country <- substr(dirn, nchar(dirn)-2, nchar(dirn))

inf <- read.csv(gsub("SNP", "variety-info", f1))
x <- matchpoint::read_dart(f1)
x2 <- matchpoint::read_dart(f2)
xc <- matchpoint::read_dart(fc)

cns <- colnames(x$snp)[-1]

stopifnot(all(colnames(x$snp)[-1] == colnames(x2$snp)[-(1:2)]))

fxl <- gsub("_SNP.csv", "_IBS.xlsx", f1)
fxl <- gsub("input", "output/IBS", fxl)
dibs <- readxl::read_excel(fxl, "distance")
bibs <- readxl::read_excel(fxl, "best_match") |> data.frame()
bibs <- bibs[!is.na(bibs$field_id), ]
```

```{r}
fpdf <- ""
if (is_html) {
   fpdf <- paste0("You can also use a <a href=", country, "_", crop, ".pdf", ">pdf version</a> of this document.")
}
```


### Introduction

`r fpdf`

Describing order: `r ordnr` for `r crop` in `r country`

The raw data had `r nrow(x$snp)` SNPs and `r ncol(x$snp)-1` samples (`r sum(inf$reference)` references and `r sum(!inf$reference)` field samples).  

```{r missing, fig.width=6, fig.height=3}
par(mfrow=c(1,2), mar=c(4,4,1,0))

snp <- x$snp[,-c(1:2)]
mr <- rowSums(is.na(snp)) / ncol(snp)
msnp <- sum(mr >.2)
msnpp <- round(100 * msnp / length(mr), 1)
snp <- snp[mr <= 0.2, ]

mrg <- colSums(is.na(snp)) / nrow(snp)

plot(sort(mr), las=1, cex=.5, col="blue", ylab="Fraction missing data", xlab="Markers", ylim=c(0,1), cex.axis=.8)
grid()

par(mar=c(4,1,1,3))
plot(sort(mrg), las=1, cex=.5, col="red", ylab="", ylim=c(0,1), xlab="Genotypes", yaxt = "n", cex.axis=.8)
grid()
```
</br></br>
`r msnp` markers (`r msnpp`%) had a missing rate that was higher than 0.2. After removing these, we had `r nrow(snp)` SNPs left.

```{r markvariation}
i <- apply(snp, 1, \(i) length(na.omit(unique(i))))
nrow1 <- nrow(snp)
snp <- snp[i>1,]

sd <- as.data.frame(table(i))
sd[,1] <- as.integer(as.character(sd[,1]))
sinf <- ifelse(nrow(snp) < nrow1, paste(nrow(snp), "of the "), "All the ")
sinfp <- round(100 * nrow(snp) / nrow(x$snp), 1)
Markers <- c("No variation", "Two variants found", "Three variants found")[sd[,1]]
sd2 <- data.frame(Markers, Count=sd$Freq)
ktab(sd2, "Variation across genotypes in SNP markers")
```

</br>
`r sinf` remaining markers were informative in the sense that they showed variation between genotypes leaving us with (`r sinfp`% ) of the original number of markers. 
</br>

```{r genotypes, fig.width=5}
mg <- sum(mrg >.2)
mgp <- round(100 * mg / length(mrg), 1)
snp <- snp[, mrg <= 0.2]
m <- match(colnames(snp), inf$sample)
nref <- sum(inf$reference[m])
nfld <- sum(!inf$reference[m])
sref <- colnames(snp) %in% inf$sample[inf$reference]

mgs <- ifelse(mg == 0, "No", mg)
```

</br>
`r mgs` genotypes (`r mgp`%) had a missing rate that was higher than 0.2. After removing these, we have `r ncol(snp)` genotypes left (`r nref` references and `r nfld` field samples).

### Single genotype?

In the case of clonally propagated or inbreeding species, one would expect that, if a sample consisted of a single genotype, the counts for each heterozygous SNP variant would have a proportion of aout 1. The "heterozygosity index" is the median score of these proportions for all heterozygous markers of a variety. Samples with less then 20% heterozygous markers are not considered.

```{r purity, fig.width=4, fig.height=4, cache=FALSE}
out <- matchpoint:::purity(xc, mr=0.2, minhet=0)  |> sort()

nv <- length(out)
par(mar=c(4,4,1,1))
plot((1:nv)/nv, out, las=1, pch=20, 
     cex=.5, lwd=2,
     xlab="fraction of samples", 
     ylab="heterozygosity index",
     col="orange")
grid()

```


### Reference sample

```{r gatherref}
ibs <- bibs[bibs$IBS > 0.8,]
bm <- as.data.frame(table(ibs$ref_id))
x2snp <- x2$snp[,-2]
x2snp <- matchpoint:::remove_sparse_records(x2snp, rows=2, verbose=F)
sref <- colnames(x2snp) %in% inf$sample[inf$reference]
dst <- matchpoint:::hamming_distance(x2snp[, sref])
#hc <- dendextend::find_dend(as.dist(dst))
hc <- hclust(as.dist(dst))
labs <- hc$labels[hc$order]
j <- match(labs, inf$sample)
k <- match(labs, bm[,1])
hits <- bm[k, 2]
hits[!is.na(hits)] <- paste0(" [", hits[!is.na(hits)], "]")
hits[is.na(hits)] <- ""
variety <- inf$variety[j]
if (!all(is.na(inf$inventory[j]))) {
	variety2 <- paste0(variety, " (", inf$sample[j], " / ", inf$inventory[j],") ", hits)
} else {
	variety2 <- paste0(variety, " (", inf$sample[j], ") ", hits)
}
vars <- na.omit(inf$variety)
vars <- vars[vars != ""]
bm <- bm[rev(order(bm$Freq)), ]
colnames(bm) <- c("variety", "number of matched field samples")
```

`r nrow(ibs)` (`r round(100*nrow(ibs)/nrow(bibs), 1)`%) of the reference samples were matched with IBS. 

The reference sample had `r length(unique(vars))` different varieties, each one represented `r round(length(vars)/length(unique(vars)), 1)` times, on average.

`r nrow(bm)` of these varieties were matched to at least one reference sample. The most commonly found variety was `r bm[1,1]` which was matched `r bm[1,2]` times. 

<div class = "row">
<div class = "col-md-6">
```{r plotmatch, fig.height=5, fig.width=5}
lab <- as.character(bm[,1])
lab[bm[,2] < 15] <- ""
par(mar=c(4, 4, 0, 1))
plot(bm[,2], las=1, cex=.4, col="red", pch=20, ylab="nuber of matches", xlab="order")
text(1:nrow(bm), bm[,2], lab, cex=.5, xpd=TRUE, pos=4, col="blue")
```
</div>
<div class = "col-md-5">
```{r tabmatch}
scrolltab(bm, "Number of matched field samples by reference variety")
```
</div></div>


```{r vars}
#diag(dst) <- NA
jj <- match(colnames(dst), inf$sample)
vars <- inf$variety[jj]
v <- matchpoint:::var_dist(dst, vars)
```

The median average Hamming distance between samples of the same variety is `r round(median(v), 1)`% mutations (the number of mutations relative to the number of usable SNPs in the assay). This is before correcting the reference for any obvious inconsistencies.

```{r bplot, fig.width=3, fig.height=3.5}
par(mar=c(4, 6, 1, 0))
fv <- 100*v/nrow(snp)
boxplot(fv, ylab="Dissimilarity within varieties\n(% mutations)", las=1)
```

## Reference scrub

There is a number of problems that can be encountered in a reference set. (1) Some varieties cannot be clearly separated, and these are combined into a single variety. These are shown below, with the original names separated by ` # `; (2) Some replicateds of a variety are clearly in the wrong group. These are renamed to the group the belong to (the renamed references get an `*` appended to their name); (3) There are multiple separate groups for the same variety. These are labeled `a`, `b`, etc (not implemented yet). 

Below is a dendrogram based on the Hamming distance and complete linkage clustering. The unit distance is the number of mutations that separate each reference. The labels are formatted as follows: `variety (sample_ID / inventory_ID) [number of IBS matches]` or `variety (sample_ID) [number of IBS matches]` if there was no inventory_ID.

The red line shows the cut off used when making new groups based on the distance matrix (not on this particular dendrogram representation). 


```{r plotdendo2, fig.width=10, fig.height=sum(sref)/8}
set.seed(1)
#rc <- sample(rainbow(length(unique(variety))))
rc <- sample(viridis::turbo(length(unique(variety))))
cols <- rep(NA, length(variety))
ref <- as.integer(as.factor(variety))
cols <- rc[ref]
dc <- as.dendrogram(hc)
dc <- dendextend::set(dc, "labels", variety2)
dc <- dendextend::set(dc, "labels_cex", 0.55)
dc <- dendextend::set(dc, "labels_col", cols)
qth <- quantile(v)
#layout(cbind(1,2), c(0.7, 0.3))
#par(mar=c(3, 1, 0, 9))
par(mar=c(3, 1, 0, 20))

plot(dendextend::highlight_branches_lwd(dc), horiz=TRUE, nodePar=list(cex=.1), cex=.6)
#lines(cbind(qth[2], c(0, length(variety))), col=gray(.5, alpha=.25), lwd=.5)
#lines(cbind(qth[3], c(0, length(variety))), col=gray(.5, alpha=.25), lwd=1)
#lines(cbind(qth[4], c(0, length(variety))), col=gray(0.25, alpha=.25), lwd=2)
#lines(cbind(qth[5], c(0, length(variety))), col=rgb(0,1,0,0.5), lwd=3)

maxcut = 250
if (country == "Tanzania") {
  if (crop=="rice") maxcut = 100
}

# refine
j <- match(colnames(dst), inf$sample)
cut <- max(15, qth[2])
ddst = dst
startvars <- inf$variety[j]
cr <- matchpoint:::clean_reference(dst, startvars, cut, nmax=100, verbose=FALSE) 

cr <- matchpoint:::clean_reference(dst, vars, min(75, max(50, qth[3])), nmax=100, verbose=FALSE) 

j <- match(colnames(dst), cr$sample)
vars <- gsub(" \\*", "", cr$to[j])
v <- matchpoint:::var_dist(dst, vars)
cut <- min(maxcut, min(150, quantile(v)[3]))
cr <- matchpoint:::clean_reference(dst, vars, cut, nmax=100, verbose=FALSE)

j <- match(colnames(dst), cr$sample)
vars <- gsub(" \\*", "", cr$to[j])
v <- matchpoint:::var_dist(dst, vars)
cut <- min(maxcut, quantile(v)[4])
cr <- matchpoint:::clean_reference(dst, vars, cut, nmax=100, verbose=FALSE) 

j <- match(labs, cr$sample)
var1 <- gsub(" \\*", "", cr$to[j])
var2 <- gsub(" \\*", "", var1)

j <- match(labs, inf$sample)
var0 <- inf$variety[j]
changed <- var0 != var2

set.seed(1)
rc2 <- sample(viridis::turbo(length(unique(var2))))
cols2 <- rep(NA, length(var2))
ref2 <- as.integer(as.factor(var2))
cols2 <- rc[ref2]

#plot(cbind(0, 1:length(var1)), type="n", axes=FALSE)
pd <- diff(par("usr")[1:2])/3.5

var1[!changed] <- ""
text(cbind(pd, 1:length(var1)), labels=var1, col=cols2, pos=4, xpd=TRUE, cex=0.55)

text(0, length(var1)+2, labels="Orignal name", pos=4, xpd=TRUE, cex=1)
text(pd, length(var1)+2, labels="Fixed name", pos=4, xpd=TRUE, cex=1)

lines(cbind(cut, c(0, length(variety))), col=rgb(1,0,0,0.5), lwd=1)
text(cut, 0, labels=round(cut), pos=1, xpd=TRUE, cex=.5, col="red")

#points(cbind(pd, 1:length(var1)), xpd=T) #set.seed(1)
#rc <- sample(rainbow(length(unique(var2))))
#cols <- rep(NA, length(var2))
#cols <- rc[as.integer(as.factor(var2))]
#text()
#par(mar=c(3, 1, 0, 10))
#plot(dendextend::highlight_branches(dc), #horiz=TRUE, nodePar=list(cex=.1), cex=.6)
#lines(cbind(cut, c(0, length(variety))), #col=gray(.5, alpha=.5), lwd=1)
```

The table below shows the changes made in the labels. 

```{r changtab}
changed <- merge(inf[, c("sample", "variety")], cr[, c("sample", "to")], by="sample")
colnames(changed)[2] <- "from"
changed <- changed[changed$from != changed$to,]
scrolltab(changed, "", width="50%")
```

</br>
</br>
