---
title: "Describe Reference"
date: "Compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    code_folding: hide
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
docache <- TRUE

is_html <- knitr::is_html_output()
knitr::opts_chunk$set(echo=is_html)
#knitr::opts_chunk$set(echo = TRUE)

this <- system("hostname", TRUE)
if (this == "LAPTOP-IVSPBGCA") { 
	dpath <- "G:/.shortcut-targets-by-id/1mfeEftF_LgRcxOT98CBIaBbYN4ZHkBr_/share/image"
} else if (this == "DESKTOP-M2BA7AA") {
	dpath <- "google drive path"
}

ktab <- function(data, caption) {
    kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width=FALSE) |> 
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))
}

scrolltab <- function(data, caption, width="100%") {
  rownames(data) <- NULL
  kableExtra::kbl(data, caption=caption) |>
    kableExtra::kable_classic(full_width = F,  position = "right") |>
    kableExtra::kable_styling(bootstrap_options = c("striped", "hover"))  |> 
    kableExtra::scroll_box(width=width, height="400px")
}
```


```{r fileread, include=FALSE, cache=docache}
ordnr <- "DRi23-7825"
#"DRi23-7955_moreOrders"

crop <- substr(ordnr, 2, 3)

f1 <- list.files(file.path(dpath, "input"), 
              pattern=paste0(ordnr, "_SNP.csv"), recursive=TRUE, full=TRUE)
stopifnot(length(f1) == 1)
f2 <- gsub("_SNP.csv$", "_SNP_2row.csv", f1)
fc <- gsub("_SNP.csv$", "_Counts.csv", f1)

dirn <- dirname(f1)
country <- substr(dirn, nchar(dirn)-2, nchar(dirn))

dart1 <- matchpoint::read_dart(f1)
dart2 <- matchpoint::read_dart(f2)
dartc <- matchpoint::read_dart(fc)
snp1 <- matchpoint:::remove_sparse_records(dart1$snp, sample_mr=0.5, snp_mr=0.5, 1, verbose=FALSE)
snp2 <- matchpoint:::remove_sparse_records(dart2$snp, sample_mr=0.5, snp_mr=0.5, 2, verbose=FALSE)
snpc <- matchpoint:::remove_sparse_records(dartc$snp, sample_mr=0.5, snp_mr=0.5, 2, verbose=FALSE)

cns1 <- colnames(dart1$snp)[-1]
inf <- read.csv(gsub("SNP", "variety-info", f1))
inf <- inf[inf$reference,]
all <- sum(inf$sample %in% colnames(dart1$snp))
removed <- all - sum((inf$sample %in% colnames(snp1)))

cn <- gsub("\\.1$", "", colnames(snp2))
i <- which(cn %in% inf$sample)
snp2 <- snp2[,i]
dst <- matchpoint:::hamming_distance(snp2)

inf <- inf[inf$sample %in% cn, ]
inf$inventory[is.na(inf$inventory)] <- inf$variety[is.na(inf$inventory)]
nvars <- length(unique(inf$variety))
tab0 <- as.data.frame(table(inf$variety))
ndups <- round(mean(tab0$Freq), 1)
tab1 <- as.data.frame(table(inf$variety, inf$inventory))
tab1 <- tab1[tab1$Freq > 0, ]
techdups <- round(mean(tab1$Freq), 2)
tab2 <- table(tab1$Var1)
biodups <- round(mean(tab2), 2)

fxl <- gsub("_SNP.csv", "_IBS.xlsx", f1)
fxl <- gsub("input", "output/IBS", fxl)
dibs <- readxl::read_excel(fxl, "distance") |> data.frame(check.names=FALSE)
rownames(dibs) <- dibs[,1]
dibs <- dibs[,-1] |> as.matrix()
if (ordnr == "DEra22-7523_1_moreOrders") {
  colnames(dibs) <- gsub("_D.$", "", colnames(dibs))
}
cn <- which(colnames(dibs) %in% inf$sample)
dibs <- dibs[cn, cn]

bibs <- readxl::read_excel(fxl, "best_match") |> data.frame()
bibs <- bibs[!is.na(bibs$IBS), ]

fdap <- gsub("IBS", "DAP", fxl)
bdap <- readxl::read_excel(fdap, "res_summary") |> data.frame()
best <- merge(bibs[, c("field_id", "IBS", "ref_id", "variety")], bdap[, c("field_id", "ref_id", "variety")], by="field_id")
best <- best[best$IBS >= 0.8, ]
colnames(best) <- gsub("\\.x$", ".IBS", colnames(best))
colnames(best) <- gsub("\\.y$", ".DAP", colnames(best))

#hits <- merge(data.frame(table(best[, "variety.x"])), data.frame(table(best[, "variety.y"])), by=1)
pdftxt <- ifelse(is_html, paste0('<a href="', country, "_", crop, '_reference.pdf">pdf version</a>'), "")
```

`r pdftxt`

### Introduction

This document describes the reference samples for `r crop` in `r country` (order `r ordnr`). There were `r sum(inf$reference)` samples `r ifelse(removed > 0, paste0("(after removing ", removed, " sample(s) for having more than 50% missing data)"), "")`, representing `r nvars` varieties. On average, there were 
`r ndups` samples for each reference variety, `r techdups` samples from the same source ("inventory"; "technical duplicates") and `r biodups` samples from different sources ("biological duplicates").  


### Identification problems

There is a number of problems that can be encountered in a reference set. 

1) Some varieties cannot be clearly separated with the SNP data. These were combined into a single variety, with the original names separated by ` # ` 

2) Some replicates of a variety are clearly in the wrong group. These were renamed to the group the belong to (the renamed references get an `*` appended to their name); 

3) There can be multiple separate groups for the same variety that are genetically very distant. These are identified by appended labels `__a`, `__b`, etc. 

Below is a dendrogram based on the Hamming distance and complete linkage clustering. The unit distance is the number of mutations that separate each reference. The labels are formatted as follows: `variety (sample_ID / inventory_ID) [number of IBS matches|number of DAP matches]`. 
The matches and inventory_ID are ommited if there were none.The red line shows the cut off used when making new groups based on the distance matrix (not on this particular dendrogram representation). 


```{r plotdendo2, fig.width=10, fig.height=sum(inf$reference)/8}
hc <- hclust(as.dist(dst))
labs <- hc$labels[hc$order]

b_ibs <- as.data.frame(table(best$ref_id.IBS))
b_dap <- as.data.frame(table(best$ref_id.DAP))
k <- match(labs, b_ibs[,1])
h_ibs <- b_ibs[k, 2]
k <- match(labs, b_dap[,1])
h_dap <- b_dap[k, 2]
h <- paste0(h_ibs, "|", h_dap)
h <- gsub("NA", "0", h)
h <- paste0("  [", h, "]")
hits <- gsub("  \\[0\\|0]", "", h)

labs <- gsub("\\.1$", "", labs)

j <- match(labs, inf$sample)

variety <- inf$variety[j]
if (!all(is.na(inf$inventory[j]))) {
	variety2 <- paste0(variety, "  (", inf$sample[j], " / ", inf$inventory[j],") ", hits)
} else {
	variety2 <- paste0(variety, "  (", inf$sample[j], ") ", hits)
}
vars <- na.omit(inf$variety)
vars <- vars[vars != ""]


set.seed(1)
#rc <- sample(rainbow(length(unique(variety))))
rc <- sample(viridis::turbo(length(unique(variety))))
cols <- rep(NA, length(variety))
ref <- as.integer(as.factor(variety))
cols <- rc[ref]
dc <- as.dendrogram(hc)
dc <- dendextend::set(dc, "labels", variety2)
dc <- dendextend::set(dc, "labels_cex", 0.55)
dc <- dendextend::set(dc, "labels_col", cols)

#layout(cbind(1,2), c(0.7, 0.3))
#par(mar=c(3, 1, 0, 9))
par(mar=c(3, 0, 0, 35))
plot(dendextend::highlight_branches_lwd(dc), horiz=TRUE, nodePar=list(cex=.1), cex=.6)

maxcut = 250
#if (country == "Tanzania") {
  if (crop=="rice") maxcut <- 50
#}

# refine
j <- match(colnames(dst), inf$sample)
startvars <- vars <- inf$variety[j]
v <- matchpoint:::var_dist(dst, vars)
qth <- quantile(v)
outh <- min(400, qth[4] + 1.5 * (qth[4] - qth[2]))
if (outh < qth[5]) {
  vars <- matchpoint:::break_far_relatives(dst, vars, outh)
  v <- matchpoint:::var_dist(dst, vars)
  qth <- quantile(v)
}
cut <- max(1, qth[2])
cr <- matchpoint:::clean_reference(dst, vars, cut, nmax=100, verbose=FALSE) 

j <- match(colnames(dst), cr$sample)
v <- matchpoint:::var_dist(dst, gsub(" \\*", "", cr$to[j]))
cut <- min(maxcut, max(2, quantile(v)[3]))
cr <- matchpoint:::clean_reference(dst, vars, cut, nmax=100, verbose=FALSE)

j <- match(colnames(dst), cr$sample)
v <- matchpoint:::var_dist(dst, gsub(" \\*", "", cr$to[j]))
cut <- min(maxcut, max(3, quantile(v)[4]))
cr <- matchpoint:::clean_reference(dst, vars, cut, nmax=100, verbose=FALSE) 

j <- match(labs, cr$sample)
#var1 <- gsub(" \\*", "", cr$to[j])
var1 <- cr$to[j]
var2 <- gsub(" \\*", "", var1)

j <- match(labs, inf$sample)
var0 <- inf$variety[j]
changed <- var0 != var2

set.seed(1)
rc2 <- sample(viridis::turbo(length(unique(var2))))
cols2 <- rep(NA, length(var2))
ref2 <- as.integer(as.factor(var2))
cols2 <- rc[ref2]

#plot(cbind(0, 1:length(var1)), type="n", axes=FALSE)
pd <- diff(par("usr")[1:2]) 

var1[!changed] <- ""
text(cbind(pd, 1:length(var1)), labels=var1, col=cols2, pos=4, xpd=TRUE, cex=0.55)

text(0, length(var1)+2, labels="Original name", pos=4, xpd=TRUE, cex=1)
text(pd, length(var1)+2, labels="Fixed name", pos=4, xpd=TRUE, cex=1)

lines(cbind(cut, c(0, length(variety))), col=rgb(1,0,0,0.5), lwd=1)
text(cut, 0, labels=round(cut), pos=1, xpd=TRUE, cex=.5, col="red")

k <- match(colnames(dibs), cr$sample)
ibsvars <- gsub(" \\*", "", cr$to[k])
vibs <- matchpoint:::var_dist(as.matrix(dibs), ibsvars)

refs <- intersect(colnames(dst), colnames(dibs))
i <- colnames(dst) %in% refs
j <- colnames(dibs) %in% refs
x <- cbind(expand.grid(refs, refs), dst=as.vector(dst[i,i]), ibs=round(1000*as.vector(dibs[j,j])))
x <- x[x[,1] != x[,2], ]
x[,1:2] <- t(apply(x[,1:2], 1, sort))
x <- unique(x)
x$ibs = x$ibs/1000
colnames(x)[1:2] <- c("id1", "id2")
inf2 <- merge(inf, cr[, c("sample", "to")], by="sample")
x <- merge(x, inf2[, c("sample", "inventory", "to")], by=1)
colnames(x)[5:6] <- c("inv1", "var1")
x <- merge(x, inf2[, c("sample", "inventory", "to")], by.x=2, by.y=1)
colnames(x)[7:8] <- c("inv2", "var2")
x <- x[x$var1 == x$var2,]

trep_ibs <- median(x$ibs[x$inv1 == x$inv2])
brep_ibs <- median(x$ibs[x$inv1 != x$inv2])
trep_dst <- median(x$dst[x$inv1 == x$inv2])
brep_dst <- median(x$dst[x$inv1 != x$inv2])

scores <- round(c(trep_dst, brep_dst, trep_ibs, brep_ibs), 3)

```

The table below shows the changes made in the labels. 

```{r changtab}
changed <- merge(inf[, c("sample", "variety")], cr[, c("sample", "to")], by="sample")
colnames(changed)[2] <- "from"
changed <- changed[changed$from != changed$to,]
scrolltab(changed, "", width="50%")
```

</br>
</br>

After the cleaning suggested above, the median average Hamming distance between samples of the same variety was `r round(median(v), 1)` mutations; `r scores[1]` for technical replicates and `r scores[2]` for biological replicates. 

The median average IBS distance between samples of the same variety was `r round(median(vibs), 3)`; `r scores[3]` for technical replicates and `r scores[4]` for biological replicates. 

```{r bplot, fig.width=6, fig.height=3.5}
j <- match(colnames(dst), cr$sample)
vars <- gsub(" \\*", "", cr$to[j])
v <- matchpoint:::var_dist(dst, vars)

par(mar=c(4, 6, 1, 0), mfrow=c(1,2))
boxplot(v, ylab="Dissimilarity within varieties\n(% mutations)", las=1)
boxplot(vibs, ylab="Dissimilarity within varieties\n(IBS)", las=1)
```


### Single genotype?

In the case of diploid clonally propagated crops and inbreeding species, one would expect that, if a sample consisted of a single genotype, the counts for each heterozygous SNP variant would have a ratio of about 1. The "heterozygosity index" is the median score of these ratios for all heterozygous markers of a variety. Samples with less then 20% heterozygous markers are not considered.

```{r purity, fig.width=4, fig.height=4, cache=FALSE}

out <- matchpoint:::purity(dartc, mr=0.5, minhet=0)  |> sort()

nv <- length(out)
par(mar=c(4,4,1,1))
plot((1:nv)/nv, out, las=1, pch=20, 
     cex=.5, lwd=2,
     xlab="fraction of samples", 
     ylab="heterozygosity index",
     col="orange")
grid()

```
